### PR TITLE
Offline-Imagery: Add Tile classes to the local db.

### PR DESCRIPTION
**Pull request by scolsen**
In order to represent what tiles users have available for offline viewing, we'll make use of the local data store. We store tile state along with the id of the json feature the tile is associated with and the path at which the tile data is stored. This should allow us to keep track of what tiles are on the device, as well as resume failed or interrupted downloads.

### CODE REVIEW

File: gnd/src/main/java/com/google/android/gnd/model/tile/Tile.java [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+import com.google.auto.value.AutoValue;
+
+import java.net.URL;
+
+@AutoValue
+public abstract class Tile {
+    public enum State {
+        PENDING,
+        DOWNLOADED,
+        FAILED
```
Comment (gino-m) [inline gnd/src/main/java/com/google/android/gnd/model/tile/Tile.java]: We may also need STARTED/IN_PROGRESS?

File: gnd/src/main/java/com/google/android/gnd/persistence/local/room/TileEntity.java [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+    public static Tile toTile(TileEntity tileEntity) {
+        Tile.Builder tile =
+                Tile.newBuilder()
+                        .setId(tileEntity.getId())
+                        .setPath(tileEntity.getPath())
+                        .setState(toTileState(tileEntity.getState()));
+        return tile.build();
+    }
+
+    private static Tile.State toTileState(TileEntityState state) {
```
Comment (gino-m) [inline gnd/src/main/java/com/google/android/gnd/persistence/local/room/TileEntity.java]: I really wish there was a better way to define and map between these representations. Maybe there's something we can do in the future with https://kotlinlang.org/docs/reference/data-classes.html

[just sharing, no action required]

File: gnd/src/main/java/com/google/android/gnd/persistence/local/room/TileEntity.java [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+        switch (state) {
+            case PENDING:
+                return Tile.State.PENDING;
+            case DOWNLOADED:
+                return Tile.State.DOWNLOADED;
+            case FAILED:
+                return Tile.State.FAILED;
+            default:
+                return Tile.State.PENDING;
```
Comment (gino-m) [inline gnd/src/main/java/com/google/android/gnd/persistence/local/room/TileEntity.java]: throw new IllegalArgumentException("Unknown tile state: " + state);

File: gnd/src/main/java/com/google/android/gnd/persistence/local/room/TileEntityState.java [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+    public static int toInt(@Nullable TileEntityState value) {
+        return IntEnum.toInt(value, UNKNOWN);
+    }
+
+    @NonNull
+    @TypeConverter
+    public static TileEntityState fromInt(int intValue) {
+        return IntEnum.fromInt(values(), intValue, UNKNOWN);
+    }
+}
```
Comment (gino-m) [inline gnd/src/main/java/com/google/android/gnd/persistence/local/room/TileEntityState.java]: Nit: Add CR

File: gnd/src/main/java/com/google/android/gnd/persistence/local/room/TileEntity.java [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+    public abstract static class Builder {
+        public abstract Builder setId(String newId);
+
+        public abstract Builder setPath(String newPath);
+
+        public abstract Builder setState(TileEntityState newState);
+
+        public abstract TileEntity build();
+    }
+}
```
Comment (gino-m) [inline gnd/src/main/java/com/google/android/gnd/persistence/local/room/TileEntity.java]: Nit: Add CR

google-java-format doesn't do this for us?

File: gnd/src/main/java/com/google/android/gnd/model/tile/Tile.java [hunk 6]
```diff
@@ -0,0 +1,29 @@
+package com.google.android.gnd.model.tile;
```
[the remark this answers is not in the document]
Comment (scolsen) [reply gnd/src/main/java/com/google/android/gnd/model/tile/Tile.java]: Good call, thanks!

File: gnd/src/main/java/com/google/android/gnd/model/tile/Tile.java (hunk 1 above)
Comment (scolsen) [reply gnd/src/main/java/com/google/android/gnd/model/tile/Tile.java]: Right. Added!

File: gnd/src/main/java/com/google/android/gnd/persistence/local/room/TileEntity.java (hunk 2 above)
Comment (scolsen) [reply gnd/src/main/java/com/google/android/gnd/persistence/local/room/TileEntity.java]: Agreed! Are we planning on rolling our own support for that, or are we going to switch to Kotlin :D?

File: gnd/src/main/java/com/google/android/gnd/persistence/local/room/TileEntity.java (hunk 5 above)
Comment (scolsen) [reply gnd/src/main/java/com/google/android/gnd/persistence/local/room/TileEntity.java]: I just ran it and nope :/

File: gnd/src/main/java/com/google/android/gnd/persistence/local/room/TileEntity.java (hunk 3 above)
Comment (scolsen) [reply gnd/src/main/java/com/google/android/gnd/persistence/local/room/TileEntity.java]: Done, makes much more sense :)

**Comment by scolsen** [comment]
@gino-m Looks like my small fixes invalidated your first review and approval :/

Mind re-approving when you get a chance?

**Pull request merged**